Fix stack traces when testing plugins

jonathanhefner opened the pull request:
`Minitest.plugin_rails_init` sets [`Minitest.backtrace_filter` to `Rails.backtrace_cleaner`](https://github.com/rails/rails/blob/16899487a4cd88b37a0a35dbbaf8c19e42cf13da/railties/lib/minitest/rails_plugin.rb#L43) right before tests are run, overwriting the value set in test_helper.rb.

`Rails.backtrace_cleaner` silences backtrace lines that do not [start with `Rails.root`](https://github.com/rails/rails/blob/16899487a4cd88b37a0a35dbbaf8c19e42cf13da/railties/lib/rails/backtrace_cleaner.rb#L13-L16) [followed by e.g. "lib/" or "test/"](https://github.com/rails/rails/blob/16899487a4cd88b37a0a35dbbaf8c19e42cf13da/railties/lib/rails/backtrace_cleaner.rb#L24).  Thus when `Rails.root` is a subdirectory of the project directory -- for example, when testing a plugin that has a dummy app -- all lines of the backtrace are silenced.

This commit adds a fallback such that when all backtrace lines are silenced, the original `Minitest.backtrace_filter` is used instead.

Additionally, this commit refactors and expands existing test coverage.

kaspth commented on railties/lib/minitest/rails_plugin.rb:
```diff
@@ -40,7 +40,14 @@ def self.plugin_rails_options(opts, options)
+        if ::Rails.root.dirname.basename.to_s == "test" && ::Rails.root.join("../../lib/").exist?
+          ::Rails.backtrace_cleaner.root = ::Rails.root.join("../../")
+        end
+      end
```
Yeah, this definitely doesn't seem right.

kaspth commented on railties/lib/rails/backtrace_cleaner.rb:
```diff
@@ -23,5 +23,11 @@ def initialize
+
+    def root=(path)
+      @root = "#{path.to_s.chomp("/")}/"
+    end
```
Yeah, also not liking that we're exposing `root`. (Both the reader/writer should be marked `# :nodoc:` by the way)

kaspth commented:
Hm, this does seem rather tricky. It seems to me that the `Rails.backtrace_cleaner` that a Rails plugin/engine returns isn't suitable for its own use, so that's where I'd look.

jonathanhefner commented:
@kaspth  Unless I'm misunderstanding what you are saying, it isn't the plugin / engine that is returning (setting) the `Rails.backtrace_cleaner`.  As best as I can tell, `Rails.backtrace_cleaner` is [automatically set](https://github.com/rails/rails/blob/master/railties/lib/rails.rb#L50-L56) by the testing dummy app, when [`require_relative "../test/dummy/config/environment"`](https://github.com/rails/rails/blob/6f98614b1fd41e9d3313d11827f9697996b3c515/railties/lib/rails/generators/rails/plugin/templates/test/test_helper.rb.tt#L4) is executed in test_helper.rb.  Then `Minitest.plugin_rails_init` [sets `Minitest.backtrace_filter` equal to `Rails.backtrace_cleaner`](https://github.com/rails/rails/blob/a974898c5d1360cd4e7ffa526823e20dcf80a91c/railties/lib/minitest/rails_plugin.rb#L43).

What if, instead of exposing `Rails.backtrace_cleaner.root=`, we implement `Rails.backtrace_cleaner=` and then modify `Rails::BacktraceCleaner#initialize` to accept a `root` arg?  Then we could add `Rails.backtrace_cleaner = Rails::BacktraceCleaner.new(File.dirname(__dir__))` to test_helper.rb, which would later be picked up by `Minitest.plugin_rails_init`.

kaspth commented:
Ah, I see. Yeah, I'd be keen to see a patch like you suggested. I think we want to subclass and name it Rails::Engine::BacktraceCleaner though. I'd also like to that to be using `ENGINE_ROOT` instead of the filename trickery and maybe put it in `rails/plugin/test.rb`?

jonathanhefner commented on railties/lib/rails.rb:
```diff
@@ -48,13 +49,11 @@ def configuration
     def backtrace_cleaner
-      @backtrace_cleaner ||= begin
-        # Relies on Active Support, so we have to lazy load to postpone definition until Active Support has been loaded
-        require "rails/backtrace_cleaner"
```
I moved this `require` to the top of the file so that the `Rails::BacktraceCleaner` constant will be loaded for test_helper.rb.  The lazy load here is [over 10 years old](https://github.com/rails/rails/commit/9cfd1d44915f4615bbb760198cd01bf4dfc69f5a#diff-4d6979942db8e18c6397d28d1574d77aR703-R704), and I don't think it's relevant anymore.

jonathanhefner commented:
@kaspth I've updated this PR with the approach I proposed.

> I think we want to subclass and name it Rails::Engine::BacktraceCleaner though.

I can add this, but I don't know that it would make the implementation cleaner than it is now.  The functionality we want is implemented by `Rails::BacktraceCleaner`.  Configuring that functionality requires passing `root` into `Rails::BacktraceCleaner` in some way.  So `Rails::BacktraceCleaner` and a `Rails::Engine::BacktraceCleaner` subclass would look the same: they would both accept a `root` value in some way.

Unless what you meant is create a `Rails::Engine::BacktraceCleaner` class, move the functionality there, and make `Rails::BacktraceCleaner` a subclass of it.  That would keep `Rails::BacktraceCleaner` itself from having to accept a `root` arg, but I'm still not sure how much that buys us.

> I'd also like to that to be using ENGINE_ROOT instead of the filename trickery and maybe put it in rails/plugin/test.rb

`ENGINE_ROOT` is [set in bin/rails](https://github.com/rails/rails/blob/667d69cc5b0527d0c3d0887cd175114e9310d519/railties/lib/rails/generators/rails/plugin/templates/bin/rails.tt#L4), which is [not generated for all plugins](https://github.com/rails/rails/blob/667d69cc5b0527d0c3d0887cd175114e9310d519/railties/lib/rails/generators/rails/plugin/plugin_generator.rb#L155).  It would also not be set when running `rake test`.  I considered adding `ENGINE_ROOT ||= File.dirname(__dir__)` to test_helper.rb as an alernative approach, but there is Rails code that behaves differently when `ENGINE_ROOT` is set ([example](https://github.com/rails/rails/blob/667d69cc5b0527d0c3d0887cd175114e9310d519/activerecord/lib/active_record/migration.rb#L616-L620)).

jonathanhefner commented:
I just pushed a new approach which does not modify `Rails::BacktraceCleaner` nor require extra configuration in test_helper.rb.

eugeneius commented on railties/lib/minitest/rails_plugin.rb:
```diff
@@ -39,8 +42,12 @@ def self.plugin_rails_options(opts, options)
-      # Plugin can run without Rails loaded, check before filtering.
-      Minitest.backtrace_filter = ::Rails.backtrace_cleaner if ::Rails.respond_to?(:backtrace_cleaner)
+      # Can use in a Rails plugin test suite or without Rails loaded.
+      if Pathname.pwd == ::Rails.try(:root)
```
It really seems like this line should be:

```suggestion
      if ::Rails.respond_to?(:backtrace_cleaner) && !defined?(ENGINE_ROOT)
```

But as you [pointed out](https://github.com/rails/rails/pull/38443#issuecomment-603489997), `ENGINE_ROOT` is lost during `rake test` because Rake's test loader [spawns a new process](https://github.com/ruby/rake/blob/v13.0.1/lib/rake/testtask.rb#L117).

I'm tempted to suggest that we accept this limitation, especially since after https://github.com/rails/rails/pull/39304 the failure mode is an unfiltered backtrace, rather than no backtrace. It's possible to run a Rails application's tests from outside its root directory, which will confuse this check; messing with behaviour for applications to improve support for `rake test` from plugins seems like the wrong tradeoff to me.

jonathanhefner replied:
> But as you [pointed out](https://github.com/rails/rails/pull/38443#issuecomment-603489997), `ENGINE_ROOT` is lost during `rake test` because Rake's test loader [spawns a new process](https://github.com/ruby/rake/blob/v13.0.1/lib/rake/testtask.rb#L117).

Unfortunately, it's not only when running `rake test`.  `ENGINE_ROOT` is not set for simple railtie plugins.  It's [set by `bin/rails`](https://github.com/rails/rails/blob/04dcf56cec38fc91f7ac6694d8a4fad66dcdf4dd/railties/lib/rails/generators/rails/plugin/templates/bin/rails.tt#L4), which is [only generated for engine plugins](https://github.com/rails/rails/blob/04dcf56cec38fc91f7ac6694d8a4fad66dcdf4dd/railties/lib/rails/generators/rails/plugin/plugin_generator.rb#L161).  We could possibly set `ENGINE_ROOT` in `bin/test` or `test_helper.rb`, but I'm not sure that would be appropriate, and it wouldn't fix the issue for existing plugins.

> I'm tempted to suggest that we accept this limitation, especially since after #39304 the failure mode is an unfiltered backtrace

True, though the backtrace will be less pretty than expected, since it will include many lines that would normally be filtered by `Minitest::BacktraceFilter`.

> It's possible to run a Rails application's tests from outside its root directory, which will confuse this check

My thinking was that `ActiveSupport::BacktraceCleaner` *might* be more desirable in this case, since it would keep backtrace paths absolute, rather than make them relative to some other directory (`Rails.root`).

With all that said, I just pushed another new approach based on 0c8df42b50764833d5f84c1cb806aa6a84c70199 that uses `Rails.backtrace_cleaner` when applicable, but otherwise uses the default `Minitest.backtrace_filter`.

eugeneius replied:
Yes! That's even better ❤️

jonathanhefner committed:
Fix backtraces for generated plugin tests

`Minitest.plugin_rails_init` sets `Minitest.backtrace_filter` to
`Rails.backtrace_cleaner` right before tests are run, overwriting the
value set in test_helper.rb.

`Rails.backtrace_cleaner` silences backtrace lines that do not start
with `Rails.root` followed by e.g. "lib/" or "test/".  Thus when
`Rails.root` is a subdirectory of the project directory -- for example,
when testing a plugin that has a dummy app -- all lines of the backtrace
are silenced.

This commit adds a fallback such that when all backtrace lines are
silenced, the original `Minitest.backtrace_filter` is used instead.

Additionally, this commit refactors and expands existing test coverage.

rafaelfranca commented on railties/lib/minitest/rails_plugin.rb:
```diff
@@ -5,6 +5,17 @@
+    end
+
+    def filter(backtrace)
+      @preferred.filter(backtrace).presence || @fallback.filter(backtrace)
```
This depends on Active Support and I'm not seeing it being required. In reality I think we should not use `presence` here.

rafaelfranca reviewed (approved):
Should we update the spec_helpers of the plugins we have inside the repo?

actionmailbox, actiontext and activestorage are using the old code.

jonathanhefner commented:
@rafaelfranca It's not required, but it does remove code, which is always nice! :smiley:
